### PR TITLE
Fix topology hints test

### PR DESCRIPTION
**Pull request by lzhecheng**
* It should check one Node in a zone instead of each Node and its fromZone.
* Check Nodes' CPUs if they are equivalen

Signed-off-by: Zhecheng Li <zhechengli@microsoft.com>

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: test/e2e/network/topology_hints.go [hunk 1]
```diff
@@ -153,7 +153,7 @@ var _ = common.SIGDescribe("Feature:Topology Hints", func() {
 				framework.Failf("Expected zone to be specified for %s node", nodeName)
 			}
 			ginkgo.By("creating a client pod for probing the service from " + fromZone)
-			podName := "curl-from-" + fromZone
+			podName := fmt.Sprintf("client%d-curl-from-%s", i, fromZone)
```
Comment (aojea) [inline test/e2e/network/topology_hints.go:156]: I think that we leaked the environment assumptions into the test, this is an example of an execution.
```
[BeforeEach] [sig-network] Feature:Topology Hints
  /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/network/topology_hints.go:47
[It] should distribute endpoints evenly
  /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/network/topology_hints.go:52
Jan 14 06:43:17.801: INFO: DaemonSet pods can't tolerate node kind-control-plane with taints [{Key:node-role.kubernetes.io/master Value: Effect:NoSchedule TimeAdded:<nil>}], skip checking this node
Jan 14 06:43:17.808: INFO: Number of nodes with available pods controlled by daemonset topology-serve-hostname: 0
Jan 14 06:43:17.808: INFO: Node kind-worker is running 0 daemon pod, expected 1
Jan 14 06:43:22.799: INFO: DaemonSet pods can't tolerate node kind-control-plane with taints [{Key:node-role.kubernetes.io/master Value: Effect:NoSchedule TimeAdded:<nil>}], skip checking this node
Jan 14 06:43:22.803: INFO: Number of nodes with available pods controlled by daemonset topology-serve-hostname: 3
Jan 14 06:43:22.803: INFO: Number of running nodes: 3, number of available pods: 3 in daemonset topology-serve-hostname
Jan 14 06:43:22.808: INFO: DaemonSet pods can't tolerate node kind-control-plane with taints [{Key:node-role.kubernetes.io/master Value: Effect:NoSchedule TimeAdded:<nil>}], skip checking this node
Jan 14 06:43:22.808: INFO: Waiting for 3 endpoints to be tracked in EndpointSlices
[1mSTEP[0m: having hints set for each endpoint
[1mSTEP[0m: keeping requests in the same zone
[1mSTEP[0m: creating a client pod for probing the service from zone-a
Jan 14 06:43:27.830: INFO: The status of Pod curl-from-zone-a is Pending, waiting for it to be Running (with Ready = true)
Jan 14 06:43:29.834: INFO: The status of Pod curl-from-zone-a is Pending, waiting for it to be Running (with Ready = true)
Jan 14 06:43:31.835: INFO: The status of Pod curl-from-zone-a is Running (Ready = true)
Jan 14 06:43:31.837: INFO: Ensuring that requests from curl-from-zone-a pod on kind-worker node stay in zone-a zone
Jan 14 06:43:36.850: INFO: Pod client logs: Fri Jan 14 06:43:29 UTC 2022
Date: Fri Jan 14 06:43:30 UTC 2022 Try: 1
topology-serve-hostname-cvlkk
Date: Fri Jan 14 06:43:31 UTC 2022 Try: 2
topology-serve-hostname-cvlkk
Date: Fri Jan 14 06:43:32 UTC 2022 Try: 3
topology-serve-hostname-cvlkk
Date: Fri Jan 14 06:43:33 UTC 2022 Try: 4
topology-serve-hostname-cvlkk
Date: Fri Jan 14 06:43:34 UTC 2022 Try: 5
topology-serve-hostname-cvlkk
Date: Fri Jan 14 06:43:35 UTC 2022 Try: 6
topology-serve-hostname-cvlkk
Date: Fri Jan 14 06:43:36 UTC 2022 Try: 7
topology-serve-hostname-cvlkk

[1mSTEP[0m: creating a client pod for probing the service from zone-b
Jan 14 06:43:36.858: INFO: The status of Pod curl-from-zone-b is Pending, waiting for it to be Running (with Ready = true)
Jan 14 06:43:38.866: INFO: The status of Pod curl-from-zone-b is Pending, waiting for it to be Running (with Ready = true)
Jan 14 06:43:40.863: INFO: The status of Pod curl-from-zone-b is Running (Ready = true)
Jan 14 06:43:40.866: INFO: Ensuring that requests from curl-from-zone-b pod on kind-worker2 node stay in zone-b zone
Jan 14 06:43:45.873: INFO: Pod client logs: Fri Jan 14 06:43:38 UTC 2022
Date: Fri Jan 14 06:43:39 UTC 2022 Try: 1
topology-serve-hostname-fx8kl
Date: Fri Jan 14 06:43:40 UTC 2022 Try: 2
topology-serve-hostname-fx8kl
Date: Fri Jan 14 06:43:41 UTC 2022 Try: 3
topology-serve-hostname-fx8kl
Date: Fri Jan 14 06:43:42 UTC 2022 Try: 4
topology-serve-hostname-fx8kl
Date: Fri Jan 14 06:43:43 UTC 2022 Try: 5
topology-serve-hostname-fx8kl
Date: Fri Jan 14 06:43:44 UTC 2022 Try: 6
topology-serve-hostname-fx8kl
Date: Fri Jan 14 06:43:45 UTC 2022 Try: 7
topology-serve-hostname-fx8kl

[1mSTEP[0m: creating a client pod for probing the service from zone-c
Jan 14 06:43:45.882: INFO: The status of Pod curl-from-zone-c is Pending, waiting for it to be Running (with Ready = true)
Jan 14 06:43:47.888: INFO: The status of Pod curl-from-zone-c is Pending, waiting for it to be Running (with Ready = true)
Jan 14 06:43:49.886: INFO: The status of Pod curl-from-zone-c is Pending, waiting for it to be Running (with Ready = true)
Jan 14 06:43:51.887: INFO: The status of Pod curl-from-zone-c is Pending, waiting for it to be Running (with Ready = true)
Jan 14 06:43:53.886: INFO: The status of Pod curl-from-zone-c is Running (Ready = true)
Jan 14 06:43:53.892: INFO: Ensuring that requests from curl-from-zone-c pod on kind-worker3 node stay in zone-c zone
Jan 14 06:43:58.900: INFO: Pod client logs: Fri Jan 14 06:43:46 UTC 2022
Date: Fri Jan 14 06:43:47 UTC 2022 Try: 1
````

If I recall correctly this loop has to iterate for one node of each zone, so it checks that for all existing zones in the cluster, the traffic remains in the zone.

Based on that, the name is correct, what is not correct is the loop, that instead of iterating over the nodeNames and break at 3, it has to iterate over the zones and pick one node of each zone, this will guarantee that the pods names are unique too

Let's wait for @robscott for confirmation

**Comment by aojea** [comment]
/ok-to-test
/kind bug
/assign @robscott @aojea 

@lzhecheng I think we don't need a release not for this change, but let's wait for Rob to reply to my comment

**Comment by lzhecheng** [comment]
@aojea sounds reasonable. Thanks!

File: test/e2e/network/topology_hints.go (hunk 1 above)
Comment (robscott) [reply test/e2e/network/topology_hints.go:156]: Yep, this test assumes it's in a cluster with a 3 nodes, each in different zones. It would make sense to update this test to support more flexible deployments and bail out/error if it is in an unsupported one.

For reference, this test currently assumes the following:
1. All nodes have equivalent allocatable CPU
2. There are 3 nodes, each in a different zone
3. The TopologyAwareHints feature gate is enabled

I don't think we have a way to verify 3, but we could add a check for 1, and refactor the test to work if the number of nodes/zones is different. We still would need nodes to exist in more than one zone for this test to work though.

Comment (lzhecheng) [reply test/e2e/network/topology_hints.go]: @aojea @robscott thank you for your suggestionsi and I have updated the code. PTAL.

File: test/e2e/network/topology_hints.go [hunk 2]
```diff
@@ -83,6 +83,7 @@ var _ = common.SIGDescribe("Feature:Topology Hints", func() {
 		framework.ExpectNoError(err, "timed out waiting for DaemonSets to be ready")
 
 		nodeNames := e2edaemonset.SchedulableNodes(c, ds)
+		e2eskipper.SkipIfNodesAllocatableCPUsNotEqual(c, nodeNames)
```
Comment (aojea) [inline test/e2e/network/topology_hints.go:86]: I prefer to not add a new functions to skipper and resource until we have more evidence that is going to have more consumers, and doesn't seem something that can be reused, at least in a short term.
We can add the equivalent code of both functions to this test, just add comments explaining why they are needed

File: test/e2e/network/topology_hints.go (hunk 1 above)
Comment (aojea) [reply test/e2e/network/topology_hints.go]: don't add new helper functions, until we have clear we can't reuse them I prefer to keep that code within the test.
Another thing is that you can do a List instead of a Get per node, it is going to be more efficiente and reduce flakiness

File: test/e2e/framework/node/resource.go [hunk 3]
```diff
@@ -583,6 +583,27 @@ func GetSchedulableClusterZones(c clientset.Interface) (sets.String, error) {
 	return zones, nil
 }
 
+func CheckNodesAllocatableCPUsEquivalent(c clientset.Interface, nodeNames []string) (bool, error) {
+	var lastNodeCPU resource.Quantity
+	for i, nodeName := range nodeNames {
+		node, err := c.CoreV1().Nodes().Get(context.TODO(), nodeName, metav1.GetOptions{})
+		if err != nil {
+			return false, fmt.Errorf("Error when getting a Node with name '%s'", nodeName)
+		}
```
Comment (aojea) [inline test/e2e/framework/node/resource.go:592]: you can do a List() and then filter out the nodes that are not in the nodeNames array

File: test/e2e/network/topology_hints.go (hunk 1 above)
Comment (lzhecheng) [reply test/e2e/network/topology_hints.go]: Makes sense. Code is updated.

File: test/e2e/network/topology_hints.go (hunk 2 above)
Comment (lzhecheng) [reply test/e2e/network/topology_hints.go]: done

File: test/e2e/framework/node/resource.go (hunk 3 above)
Comment (lzhecheng) [reply test/e2e/framework/node/resource.go]: done

File: test/e2e/network/topology_hints.go [hunk 4]
```diff
@@ -83,6 +85,33 @@ var _ = common.SIGDescribe("Feature:Topology Hints", func() {
...
+			if !found {
+				framework.Failf("Error when getting allocatable CPU of Node '%s'", node.Name)
+			}
+			if firstNode {
+				lastNodeCPU = nodeCPU
+			} else if !nodeCPU.Equal(lastNodeCPU) {
+				skipper.Skipf("Not all Nodes have same allocatable CPUs. Value of Node '%s' is different from the last one. %d not equals %d",
+					node.Name, nodeCPU.Value(), lastNodeCPU.Value())
+			}
+			firstNode = false
```
Comment (aojea) [inline test/e2e/network/topology_hints.go:112]:
```suggestion
			if firstNode {
				lastNodeCPU = nodeCPU
				firstNode = false
			} else if !nodeCPU.Equal(lastNodeCPU) {
				skipper.Skipf("Not all Nodes have same allocatable CPUs. Value of Node '%s' is different from the last one. %d not equals %d",
					node.Name, nodeCPU.Value(), lastNodeCPU.Value())
			}
```

Comment (aojea) [reply test/e2e/network/topology_hints.go:112]: you have to switch `firstNode` to false after you set lastNodeCPU, right?

File: test/e2e/network/topology_hints.go [hunk 5]
```diff
@@ -143,15 +172,9 @@ var _ = common.SIGDescribe("Feature:Topology Hints", func() {
...
-			// Iterate through max of 3 nodes
-			if i > 2 {
-				break
-			}
-			fromZone, ok := nodesByZone[nodeName]
-			if !ok {
-				framework.Failf("Expected zone to be specified for %s node", nodeName)
-			}
+		for fromZone, nodes := range zonesWithNodes {
+			nodeName := nodes[0]
```
Comment (aojea) [inline test/e2e/network/topology_hints.go:176]: I didn't check in detail, do we guarantee that we always have at least one node per zone? otherwise this will panic

File: test/e2e/network/topology_hints.go (hunk 4 above)
Comment (lzhecheng) [reply test/e2e/network/topology_hints.go:112]: I put it at the end of the for loop. But yes, your suggestion makes it a bit faster.

File: test/e2e/network/topology_hints.go (hunk 5 above)
Comment (lzhecheng) [reply test/e2e/network/topology_hints.go:176]: I don't think panic will happen. `zonesWithNodes` is set on L161. Each zone should map to a list with at least one Node.

File: test/e2e/network/topology_hints.go [hunk 6]
```diff
@@ -83,6 +85,33 @@ var _ = common.SIGDescribe("Feature:Topology Hints", func() {
...
+		framework.ExpectNoError(err, "Error when listing all Nodes")
+		var lastNodeCPU resource.Quantity
+		firstNode := true
+		for _, node := range nodeList.Items {
+			if !nodeNamesMap[node.Name] {
+				continue
+			}
+			nodeCPU, found := node.Status.Allocatable[v1.ResourceCPU]
+			if !found {
+				framework.Failf("Error when getting allocatable CPU of Node '%s'", node.Name)
```
Comment (aojea) [inline test/e2e/network/topology_hints.go:104]: @robscott fail or skip?

**Comment by aojea** [comment]
/test pull-kubernetes-e2e-kind-multizone
Thanks
lgtm, wait for Robs and CI feedback

**Comment by lzhecheng** [comment]
/test pull-kubernetes-e2e-kind-multizone
I pushed again because the old code change included 2 `import k8s.io/kubernetes/test/e2e/framework/skipper` by mistake. I removed one this time. Last time all tests are successful except the import check.

**Comment by aojea** [comment]
/retest
test is working
LGTM

**Comment by aojea** [comment]
/approve
defer LGTM to rob

File: test/e2e/network/topology_hints.go [hunk 7]
```diff
@@ -124,12 +152,12 @@ var _ = common.SIGDescribe("Feature:Topology Hints", func() {
 			}
 		}
 
-		nodeList, err := c.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{})
-		framework.ExpectNoError(err)
 		nodesByZone := map[string]string{}
+		zonesWithNodes := map[string][]string{}
```
Comment (robscott) [inline test/e2e/network/topology_hints.go:156]: Seems like we all we really need is one node name per zone, so a simpler data structure like `nodeForZone := map[zone]nodeName{}` would work. (Where zone and node name would just be strings).

File: test/e2e/network/topology_hints.go [hunk 8]
```diff
@@ -124,12 +152,12 @@ var _ = common.SIGDescribe("Feature:Topology Hints", func() {
 			}
 		}
 
-		nodeList, err := c.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{})
-		framework.ExpectNoError(err)
 		nodesByZone := map[string]string{}
```
Comment (robscott) [inline test/e2e/network/topology_hints.go:155]: I don't think we need this any more

File: test/e2e/network/topology_hints.go (hunk 6 above)
Comment (robscott) [reply test/e2e/network/topology_hints.go:103]: Fail seems appropriate. If the test has not been skipped already, we're likely intentionally trying to run it, so failing will likely provide more noticeable/meaningful output.

File: test/e2e/network/topology_hints.go [hunk 9]
```diff
@@ -83,6 +84,33 @@ var _ = common.SIGDescribe("Feature:Topology Hints", func() {
 		framework.ExpectNoError(err, "timed out waiting for DaemonSets to be ready")
 
 		nodeNames := e2edaemonset.SchedulableNodes(c, ds)
+
+		// All Nodes should have same allocatable CPUs. If not, then skip the test.
+		nodeNamesMap := map[string]bool{}
```
Comment (robscott) [inline test/e2e/network/topology_hints.go:89]: Maybe a slight rename to indicate that this was specific to schedulable Nodes? Not tied to this specific name, but would help make the rest of the code a bit easier to understand. 
```suggestion
		schedulableNodes := map[string]bool{}
```

File: test/e2e/network/topology_hints.go [hunk 10]
```diff
@@ -83,6 +84,33 @@ var _ = common.SIGDescribe("Feature:Topology Hints", func() {
...
+			}
+			nodeCPU, found := node.Status.Allocatable[v1.ResourceCPU]
+			if !found {
+				framework.Failf("Error when getting allocatable CPU of Node '%s'", node.Name)
+			}
+			if firstNode {
+				lastNodeCPU = nodeCPU
+				firstNode = false
+			} else if !nodeCPU.Equal(lastNodeCPU) {
+				e2eskipper.Skipf("Not all Nodes have same allocatable CPUs. Value of Node '%s' is different from the last one. %d not equals %d",
```
Comment (robscott) [inline test/e2e/network/topology_hints.go:109]: Maybe reword this to clarify what we expected here.
```suggestion
				e2eskipper.Skipf("Expected Nodes to have equivalent allocatable CPUs, but Node '%s' is different from the previous one. %d not equals %d",
```

File: test/e2e/network/topology_hints.go (hunk 5 above)
Comment (robscott) [reply test/e2e/network/topology_hints.go:175]: If we move to `map[string]string{}` as I suggested above, I think this will be a non issue.

**Review by robscott — COMMENTED** [review]
Thanks for your work on this @lzhecheng!

File: test/e2e/network/topology_hints.go (hunk 5 above)
Comment (lzhecheng) [reply test/e2e/network/topology_hints.go]: done

File: test/e2e/network/topology_hints.go (hunk 7 above)
Comment (lzhecheng) [reply test/e2e/network/topology_hints.go]: Good idea!

File: test/e2e/network/topology_hints.go (hunk 8 above)
Comment (lzhecheng) [reply test/e2e/network/topology_hints.go:157]: Actually I think it is needed because `podsByZone` still needs it to set up.

File: test/e2e/network/topology_hints.go (hunk 9 above)
Comment (lzhecheng) [reply test/e2e/network/topology_hints.go]: done

File: test/e2e/network/topology_hints.go (hunk 10 above)
Comment (lzhecheng) [reply test/e2e/network/topology_hints.go]: done

**Comment by robscott** [comment]
Thanks @lzhecheng!

/lgtm
